### PR TITLE
[improve][broker] Make CompactedTopicImpl.findStartPointLoop work more efficiently

### PR DESCRIPTION
**Pull request by poorbarcode**
### Motivation & Modifications

`CompactedTopicImpl.findStartPointLoop` uses the binary search, but the implementation had a flaw that caused several more loops to be executed. If an array like this: `[1,2,3...100]`, and we search value `2` in this array, `CompactedTopicImpl.findStartPointLoop` will execute the following loops:

```
start: 0, mid:50, end: 100
start: 0, mid:25, end: 50
start: 0, mid:12, end: 25
start: 0, mid:6, end: 12
start: 0, mid:3, end: 6
start: 0, mid:1, end: 3
start: 1, mid:1, end: 1    bingo!
```

We can optimize the loop like this:

```
start: 0, mid:50, end: 100
start: 1, mid:25, end: 50    bingo!
```

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: 

- https://github.com/poorbarcode/pulsar/pull/20

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import org.testng.annotations.Test;
+
+@Test(groups = "broker-compaction")
+public class CompactedTopicImplTest {
+
+    private static final long DEFAULT_LEDGER_ID = 1;
+
+    // Sparse ledger makes multi entry has same data, this is used to construct complex environments to verify that the
+    // smallest position with the correct data is found.
+    private static final TreeMap<Long,Long> ORIGIN_SPARSE_LEDGER = new TreeMap<>();
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java:43]:
```suggestion
    private static final TreeMap<Long, Long> ORIGIN_SPARSE_LEDGER = new TreeMap<>();
```

File: pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    static {
+        ORIGIN_SPARSE_LEDGER.put(0L,0L);
+        ORIGIN_SPARSE_LEDGER.put(1L,1L);
+        ORIGIN_SPARSE_LEDGER.put(2L,1001L);
+        ORIGIN_SPARSE_LEDGER.put(3L,1002L);
+        ORIGIN_SPARSE_LEDGER.put(4L,1003L);
+        ORIGIN_SPARSE_LEDGER.put(10L,1010L);
+        ORIGIN_SPARSE_LEDGER.put(20L,1020L);
+        ORIGIN_SPARSE_LEDGER.put(50L,1050L);
+        ORIGIN_SPARSE_LEDGER.put(Long.MAX_VALUE,Long.MAX_VALUE);
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java:54]:
```suggestion
        ORIGIN_SPARSE_LEDGER.put(Long.MAX_VALUE, Long.MAX_VALUE);
```

File: pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            {3, 40, 1003},// second value & find in a range.
+            {3, 40, 1010},// not first value & find in a range.
+            {3, 40, 1015}// value not exists & find in a range.
+        };
+    }
+
+    private static CacheLoader<Long, MessageIdData> mockCacheLoader(long start, long end, final long targetMessageId,
+                                                                    AtomicLong bingoMarker){
+        // Mock ledger.
+        final TreeMap<Long,Long> sparseLedger = new TreeMap<>();
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java:75]:
```suggestion
        final TreeMap<Long, Long> sparseLedger = new TreeMap<>();
```

File: pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        sparseLedger.putAll(ORIGIN_SPARSE_LEDGER.subMap(start, end + 1));
+        sparseLedger.put(Long.MAX_VALUE,Long.MAX_VALUE);
+
+        Function<Long,Long> findMessageIdFunc = entryId -> sparseLedger.ceilingEntry(entryId).getValue();
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java:79]:
```suggestion
        Function<Long, Long> findMessageIdFunc = entryId -> sparseLedger.ceilingEntry(entryId).getValue();
```

**Review by codelipenghui — COMMENTED** [review]
Great catch!

The fix looks good to me. How can the test cover the new improvement?

File: pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java (hunk 4 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java]: Already fixed

Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java]: Already fixed

File: pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java (hunk 3 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java]: Already fixed

File: pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java (hunk 2 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java]: Already fixed

File: pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java (hunk 1 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java]: Already fixed

**Comment by poorbarcode** [comment]
Hi @codelipenghui 

> How can the test cover the new improvement?

already append test "testRecursionNumberOfFindStartPointLoop" to cover this case.

File: pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        CompactedTopicImpl.findStartPointLoop(targetPosition, start, end, promise, cache);
+        long result = promise.join();
+        assertEquals(result, bingoMarker.get());
+    }
+
+    /**
+     * Why should we check the recursion number of "findStartPointLoop", see: #17976
+     */
+    @Test
+    public void testRecursionNumberOfFindStartPointLoop() throws Exception {
```
Comment (labuladong) [inline pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java:121]: Seems that this function won't throw any `Exceptions`.

**Review by labuladong — APPROVED** [review]
Looks great to me.

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java (hunk 5 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java]: Already removed. Thanks

**Review by Technoboy- — APPROVED** [review]

**Review by coderzc — APPROVED** [review]
LGTM

**Pull request merged**